### DISCUSSION TITLE
viterbi() - Probability of each behavioral state

### DISCUSSION THREAD

**Discussion by leclermt** [bmcclintock/momentuHMM#129] [Q&A] [answered]
Hello everyone,

I was curious to know if it's possible to extract the probability of a step to be in each behavioral state with viterbi()?
The viterbi() now give the most likely behavioral state, but not the probability associated to it.
I have active and inactive behavioral state in my model.
I am ok with having inactive steps in the active behavioral state, but I am more sensible to contamination of active steps within the inactive behavioral state.
Therefore, I would like to play with the % of the most likely steps.
E.g. behavioral state = active, unless inactive probability > 75%. 
 
One alternative I thought of would be to run a model with 3 behavioral states, consequently having a “less variable” inactive behavioral state.

Thanks,
Martin

**Comment by bmcclintock** [comment, marked as answer]
If I understand correctly, I think you're looking for the state probabilities that are estimated using the `stateProbs` function. See `?stateProbs` and examples in the `momentuHMM` [vignette](https://cran.r-project.org/web/packages/momentuHMM/vignettes/momentuHMM.pdf).

**Reply by leclermt** [reply]
Awesome thanks!

**Reply by storriel** [reply]
@bmcclintock I am also planning to use the state probabilities to remove the less certain states, but for a small number of my locations (603/23514), the `viterbi` function gives a different state than that with the highest probability from the `stateProbs` function. For some of these the probability was high (up to 0.872), so I am unsure whether to trust the output of `viterbi`  or `stateProbs`. Any thoughts on this would be very much appreciated.

All the best,
Luke 

The code below starts with a model, m1, from `fitHMM`


# Decode states and compute state probabilities 
> m1_states<-viterbi(m1)
> m1_stateprobs<-stateProbs(m1)

# Combine in data frame
> m1_full<-data.frame(m1_states,m1_stateprobs)
> m1_full[,2:4]<-round(m1_full[,2:4], 3)

> head(m1_full,20)

        m1_states state.1 state.2 state.3
    1          1   1.000   0.000   0.000
    2          3   0.090   0.112   0.798
    3          3   0.017   0.095   0.888
    4          3   0.030   0.079   0.891
    5          3   0.014   0.111   0.876
    6          3   0.019   0.163   0.818
    7          3   0.077   0.010   0.913
    8          2   0.020   0.521   0.458
    9          2   0.003   0.876   0.121
    10         2   0.000   0.967   0.032
    11         2   0.000   0.996   0.004
    12         2   0.000   0.999   0.001
    13         2   0.000   0.999   0.001
    14         2   0.000   0.999   0.001
    15         2   0.000   0.996   0.003
    16         2   0.000   0.997   0.003
    17         2   0.000   0.982   0.018
    18         2   0.000   0.988   0.012
    19         2   0.001   0.850   0.149
    20         2   0.001   0.632   0.366

# Identify whether Viterbi state = state with greatest probability
> m1_full$match<-ifelse(m1_full$m1_states == '1' & m1_full$state.1 > m1_full$state.2 & m1_full$state.1 > m1_full$state.3, 'YES', 'NO')
> m1_full$match<-ifelse(m1_full$m1_states == '2' & m1_full$state.2 > m1_full$state.1 & m1_full$state.2 > m1_full$state.3, 'YES', m1_full$match)
> m1_full$match<-ifelse(m1_full$m1_states == '3' & m1_full$state.3 > m1_full$state.1 & m1_full$state.3 > m1_full$state.2, 'YES', m1_full$match)

> m1_full_nomatch<-m1_full[m1_full$match == 'NO',]

# View top 10 highest probabilities which do not match for each state
> m1_full_nomatch<-m1_full_nomatch[order(-m1_full_nomatch$state.1),]

> head(m1_full_nomatch,10)

          m1_states state.1 state.2 state.3 match
    10689         3   0.782   0.000   0.218    NO
    3936          3   0.773   0.000   0.227    NO
    7814          3   0.769   0.000   0.231    NO
    20219         3   0.754   0.000   0.246    NO
    10781         3   0.722   0.000   0.278    NO
    21975         3   0.719   0.000   0.281    NO
    1173          3   0.718   0.008   0.274    NO
    5547          3   0.704   0.000   0.296    NO
    19106         3   0.698   0.010   0.293    NO
    15995         3   0.696   0.000   0.304    NO

> m1_full_nomatch<-m1_full_nomatch[order(-m1_full_nomatch$state.2),]
> head(m1_full_nomatch,10)

          m1_states state.1 state.2 state.3 match
    2378          3   0.001   0.872   0.128    NO
    2377          3   0.001   0.868   0.131    NO
    2376          3   0.001   0.832   0.167    NO
    2379          3   0.002   0.830   0.168    NO
    2380          3   0.001   0.816   0.183    NO
    16931         3   0.001   0.781   0.218    NO
    20513         3   0.001   0.770   0.229    NO
    16932         3   0.001   0.752   0.246    NO
    20512         3   0.003   0.740   0.257    NO
    17164         3   0.001   0.735   0.264    NO

> m1_full_nomatch<-m1_full_nomatch[order(-m1_full_nomatch$state.3),]
> head(m1_full_nomatch,10)

          m1_states state.1 state.2 state.3 match
    5313          1   0.180   0.004   0.816    NO
    19439         2   0.002   0.206   0.792    NO
    2267          1   0.223   0.003   0.773    NO
    10826         1   0.234   0.000   0.766    NO
    8485          1   0.233   0.000   0.766    NO
    5314          1   0.236   0.000   0.763    NO
    20656         2   0.003   0.237   0.760    NO
    8484          1   0.239   0.000   0.760    NO
    19440         2   0.005   0.238   0.757    NO
    20657         2   0.005   0.243   0.751    NO

**Comment by bmcclintock** [comment]
The Viterbi algorithm identifies the singe most likely state sequence (i.e. "global" decoding). The forward-backward algorithm used in `stateProbs` calculates the state probabilities for each time step (i.e. "local" decoding"). The Viterbi-decoded state will thus not always correspond to the highest state probability for a given time step. It would perhaps be more concerning if the Viterbi-decoded state had an estimated state probability of 0, but that does not appear to be the case here. For more on this, see [Zucchini et al. 2016](https://www.taylorfrancis.com/books/mono/10.1201/b20790/hidden-markov-models-time-series-walter-zucchini-iain-macdonald-roland-langrock).
